[clang] Clear `NeedsCleaning` flag after `ExpandBuiltinMacro`

mariusdr opened the pull request:
After builtin macro expansion in `Preprocessor::ExpandBuiltinMacro` the result token may have the `Token::NeedsCleaning` flag set which causes an assertion failure later on when the lexer retrieves the spelling of the token in `getSpellingSlow`.

This commit adds an `Tok.clearFlag(Token::NeedsCleaning)` call to the end of `ExpandBuiltinMacro`.

Closes #128384

mariusdr committed:
[clang] Clear `NeedsCleaning` flag after `ExpandBuiltinMacro`

After builtin macro expansion in `Preprocessor::ExpandBuiltinMacro`
the result token may have the `Token::NeedsCleaning` flag set
which causes an assertion failure later on when the lexer retrives
the spelling of the token in `getSpellingSlow` in Lexer.cpp.

AaronBallman commented on clang/lib/Lex/PPMacroExpansion.cpp:
```diff
@@ -1646,7 +1646,6 @@ void Preprocessor::ExpandBuiltinMacro(Token &Tok) {
 
   // Set up the return result.
   Tok.setIdentifierInfo(nullptr);
-  Tok.clearFlag(Token::NeedsCleaning);
```
There are early returns between where we used to clear the needs cleaning flag and where we set it now; doesn't this mean we'll leave the flag in "needs cleaning" state in circumstances where it won't need cleaning (e.g., when dealing with `__DATE__`)?

mariusdr replied:
True, my bad. If the flag is removed we can get asserts on __DATE__ and __TIME__ expansion. I've put it back and added some test cases to account for this (also for some similar macros like __LINE__, __FILE__ which do not early return but are easily tested in the same way).

mariusdr committed:
Blunder keep clearFlag(Token::NeedsCleaning)

at the beginning of ExpandBuiltinMacro to account for early returns
(e.g..in __DATE__ or __TIME__ expansion).

AaronBallman reviewed (commented):
This generally LGTM, though please add a release note to `clang/docs/ReleaseNotes.rst`. I'd appreciate @cor3ntin taking a look as well before landing, just so we get another set of eyes on the changes.

mariusdr committed:
add release notes

mariusdr commented:
> This generally LGTM, though please add a release note to `clang/docs/ReleaseNotes.rst`. I'd appreciate @cor3ntin taking a look as well before landing, just so we get another set of eyes on the changes.

Thanks, added release notes under "Bug Fixes in this Version".

cor3ntin reviewed (approved):
LGTM - sorry for the delayed response.


Will you need me to merge that for you (once the merge conflict is resolved)?

mariusdr committed:
Merge branch 'main' into clang_lexer_clear_needscleaning_flag_after_expand_builtin_macro

mariusdr commented:
> LGTM - sorry for the delayed response.
> 
> Will you need me to merge that for you (once the merge conflict is resolved)?

Yes, thanks!
